### ISSUE TITLE
Possible memory leak in TimePicker

### ISSUE THREAD

**Issue by archie94** [wdullaer/MaterialDateTimePicker#587]
LeakCanary Log:

```
In org.havenapp.main:0.2.0-beta-3:214.
* EXCLUDED LEAK.
* com.wdullaer.materialdatetimepicker.time.TimePickerDialog has leaked:
* thread HandlerThread.!(<Java Local>)! (named 'queued-work-looper')
* ↳ Message.!(obj)! , matching exclusion field android.os.Message#obj
* ↳ TimePickerDialog

* Reference Key: b49afc33-9406-4be3-8268-38588066c8a8
* Device: Xiaomi Xiaomi POCO F1 beryllium
* Android Version: 9 API: 28 LeakCanary: 1.6.3 31007b4
* Durations: watch=127820ms, gc=221ms, heap dump=5780ms, analysis=15693ms

* Details:
* Instance of android.os.HandlerThread
|   static $class$clinitThreadId = 0
|   static $class$name = "android.os.HandlerThread"
|   static $class$ifTable = java.lang.Object[2]@1889701000 (0x70a28c88)
|   static $class$status = -536870912
|   static $class$primitiveType = 131072
|   static $class$classFlags = 0
|   static $class$dexCache = java.lang.DexCache@1889632912 (0x70a18290)
|   static $class$iFields = 1893732208
|   static $class$referenceInstanceOffsets = -1073741824
|   static $class$numReferenceStaticFields = 0
|   static $class$extData = null
|   static $class$objectSize = 148
|   static $class$sFields = 0
|   static $class$methods = 1895053964
|   static $class$objectSizeAllocFastPath = 152
|   static $class$shadow$_klass_ = java.lang.Class
|   static $class$shadow$_monitor_ = 536870912
|   static $class$classLoader = null
|   static $class$dexClassDefIndex = 200
|   static $class$copiedMethodsOffset = 9
|   static $class$classSize = 332
|   static $class$virtualMethodsOffset = 2
|   static $class$numReferenceInstanceFields = 2
|   static $classOverhead = byte[208]@1891144313 (0x70b89279)
|   static $class$accessFlags = 524289
|   static $class$componentType = null
|   static $class$vtable = null
|   static $class$dexTypeIndex = 866
|   static $class$superClass = java.lang.Thread
|   mHandler = null
|   mLooper = android.os.Looper@319294464 (0x13080c00)
|   mPriority = -2
|   mTid = 14465
|   blocker = null
|   blockerLock = java.lang.Object@322553416 (0x1339c648)
|   contextClassLoader = dalvik.system.PathClassLoader@319468136 (0x130ab268)
|   daemon = false
|   eetop = 0
|   group = java.lang.ThreadGroup@1883415488 (0x7042a3c0)
|   inheritableThreadLocals = null
|   inheritedAccessControlContext = java.security.AccessControlContext@322553424 (0x1339c650)
|   lock = java.lang.Object@322553432 (0x1339c658)
|   name = "queued-work-looper"
|   nativeParkEventPointer = 0
|   nativePeer = 3799507456
|   parkBlocker = null
|   parkState = 1
|   priority = 5
|   single_step = false
|   stackSize = 0
|   started = true
|   stillborn = false
|   target = null
|   threadLocalRandomProbe = 0
|   threadLocalRandomSecondarySeed = 0
|   threadLocalRandomSeed = 0
|   threadLocals = java.lang.ThreadLocal$ThreadLocalMap@322553440 (0x1339c660)
|   threadQ = null
|   threadStatus = 0
|   tid = 21282
|   uncaughtExceptionHandler = null
|   shadow$_klass_ = android.os.HandlerThread
|   shadow$_monitor_ = 0
* Instance of android.os.Message
|   static $class$clinitThreadId = 0
|   static $class$name = "android.os.Message"
|   static FLAGS_TO_CLEAR_ON_COPY_FROM = 1
|   static gCheckRecycle = true
|   static $class$ifTable = java.lang.Object[2]@1889703032 (0x70a29478)
|   static $class$status = -536870912
|   static $class$primitiveType = 131072
|   static $class$classFlags = 0
|   static $class$dexCache = java.lang.DexCache@1889632912 (0x70a18290)
|   static sPoolSync = java.lang.Object@1889633120 (0x70a18360)
|   static CREATOR = android.os.Message$1@1889703056 (0x70a29490)
|   static sPool = android.os.Message@319303208 (0x13082e28)
|   static $class$iFields = 1893729992
|   static $class$referenceInstanceOffsets = 127
|   static $class$numReferenceStaticFields = 3
|   static $class$extData = null
|   static $class$objectSize = 64
|   static $class$sFields = 1893729844
|   static FLAG_ASYNCHRONOUS = 2
|   static $class$methods = 1895048228
|   static $class$objectSizeAllocFastPath = 64
|   static $class$shadow$_klass_ = java.lang.Class
|   static FLAG_IN_USE = 1
|   static sPoolSize = 40
|   static $class$shadow$_monitor_ = 536870912
|   static MAX_POOL_SIZE = 50
|   static $class$classLoader = null
|   static $class$dexClassDefIndex = 3157
|   static $class$copiedMethodsOffset = 35
|   static $class$classSize = 289
|   static $class$virtualMethodsOffset = 13
|   static $class$numReferenceInstanceFields = 7
|   static $classOverhead = byte[132]@1889630825 (0x70a17a69)
|   static $class$accessFlags = 524305
|   static $class$componentType = null
|   static $class$vtable = null
|   static $class$dexTypeIndex = 1012
|   static $class$superClass = java.lang.Object
|   arg1 = 0
|   arg2 = 0
|   callback = null
|   data = null
|   flags = 0
|   monitorInfo = android.os.BaseLooper$MessageMonitorInfo@322356624 (0x1336c590)
|   next = null
|   obj = com.wdullaer.materialdatetimepicker.time.TimePickerDialog@322356712 (0x1336c5e8)
|   replyTo = null
|   sendingUid = -1
|   target = android.app.Dialog$ListenersHandler@322357064 (0x1336c748)
|   what = 68
|   when = 0
|   shadow$_klass_ = android.os.Message
|   shadow$_monitor_ = 0
* Instance of com.wdullaer.materialdatetimepicker.time.TimePickerDialog
|   static $class$name = null
|   static KEY_CURRENT_ITEM_SHOWING = "current_item_showing"
|   static PM = 1
|   static $class$ifTable = java.lang.Object[18]@321116960 (0x1323db20)
|   static $class$primitiveType = 131072
|   static KEY_TIMEPOINTLIMITER = "timepoint_limiter"
|   static KEY_VERSION = "version"
|   static $class$iFields = 3977096856
|   static $class$referenceInstanceOffsets = -1073741824
|   static SECOND_INDEX = 2
|   static KEY_TYPED_TIMES = "typed_times"
|   static KEY_IN_KB_MODE = "in_kb_mode"
|   static KEY_CANCEL_COLOR = "cancel_color"
|   static $class$numReferenceStaticFields = 23
|   static $class$extData = null
|   static KEY_OK_COLOR = "ok_color"
|   static $class$objectSize = 346
|   static $class$sFields = 3977096384
|   static $class$objectSizeAllocFastPath = 352
|   static KEY_THEME_DARK_CHANGED = "theme_dark_changed"
|   static KEY_DISMISS = "dismiss"
|   static $class$classLoader = dalvik.system.PathClassLoader@319468136 (0x130ab268)
|   static AM = 0
|   static $class$copiedMethodsOffset = 93
|   static $class$classSize = 1228
|   static $class$virtualMethodsOffset = 29
|   static KEY_ENABLE_SECONDS = "enable_seconds"
|   static HOUR_INDEX = 0
|   static $class$numReferenceInstanceFields = 36
|   static $class$accessFlags = 524289
|   static $class$componentType = null
|   static $class$vtable = null
|   static KEY_CANCEL_STRING = "cancel_string"
|   static $class$clinitThreadId = 14319
|   static KEY_ENABLE_MINUTES = "enable_minutes"
|   static KEY_LOCALE = "locale"
|   static $class$status = -536870912
|   static TAG = "TimePickerDialog"
|   static $class$classFlags = 0
|   static $class$dexCache = java.lang.DexCache@320318880 (0x1317ada0)
|   static KEY_THEME_DARK = "theme_dark"
|   static KEY_INITIAL_TIME = "initial_time"
|   static KEY_TITLE = "dialog_title"
|   static KEY_VIBRATE = "vibrate"
|   static MINUTE_INDEX = 1
|   static KEY_ACCENT = "accent"
|   static $class$methods = 3977097744
|   static $class$shadow$_klass_ = java.lang.Class
|   static $class$shadow$_monitor_ = 0
|   static $class$dexClassDefIndex = 7281
|   static PULSE_ANIMATOR_DELAY = 300
|   static KEY_IS_24_HOUR_VIEW = "is_24_hour_view"
|   static KEY_CANCEL_RESID = "cancel_resid"
|   static $classOverhead = byte[988]@320439609 (0x13198539)
|   static KEY_OK_RESID = "ok_resid"
|   static KEY_OK_STRING = "ok_string"
|   static $class$dexTypeIndex = 1506
|   static $class$superClass = androidx.fragment.app.DialogFragment
|   mAccentColor = -671191
|   mAllowAutoAdvance = true
|   mAmKeyCode = -1
|   mAmPmLayout = android.widget.LinearLayout@322357120 (0x1336c780)
|   mAmText = "AM"
|   mAmTextView = androidx.appcompat.widget.AppCompatTextView@322357808 (0x1336ca30)
|   mCallback = org.havenapp.main.MonitorActivity@322358648 (0x1336cd78)
|   mCancelButton = androidx.appcompat.widget.AppCompatButton@322358984 (0x1336cec8)
|   mCancelColor = -1
|   mCancelResid = 2131823302
|   mCancelString = null
|   mDefaultLimiter = com.wdullaer.materialdatetimepicker.time.DefaultTimepointLimiter@322359816 (0x1336d208)
|   mDeletedKeyFormat = "%1$s deleted"
|   mDismissOnPause = false
|   mDoublePlaceholderText = "00"
|   mEnableMinutes = true
|   mEnableSeconds = true
|   mHapticFeedbackController = com.wdullaer.materialdatetimepicker.HapticFeedbackController@322359848 (0x1336d228)
|   mHourPickerDescription = "Hours circular slider"
|   mHourSpaceView = androidx.appcompat.widget.AppCompatTextView@322359880 (0x1336d248)
|   mHourView = com.wdullaer.materialdatetimepicker.AccessibleTextView@322360720 (0x1336d590)
|   mInKbMode = false
|   mInitialTime = com.wdullaer.materialdatetimepicker.time.Timepoint@322361544 (0x1336d8c8)
|   mIs24HourMode = true
|   mLegalTimesTree = com.wdullaer.materialdatetimepicker.time.TimePickerDialog$Node@322361568 (0x1336d8e0)
|   mLimiter = com.wdullaer.materialdatetimepicker.time.DefaultTimepointLimiter@322359816 (0x1336d208)
|   mLocale = java.util.Locale@2016924456 (0x7837d328)
|   mMinutePickerDescription = "Minutes circular slider"
|   mMinuteSpaceView = androidx.appcompat.widget.AppCompatTextView@322361584 (0x1336d8f0)
|   mMinuteView = com.wdullaer.materialdatetimepicker.AccessibleTextView@322362424 (0x1336dc38)
|   mOkButton = androidx.appcompat.widget.AppCompatButton@322363248 (0x1336df70)
|   mOkColor = -1
|   mOkResid = 2131823319
|   mOkString = null
|   mOnCancelListener = null
|   mOnDismissListener = null
|   mPlaceholderText = 0
|   mPmKeyCode = -1
|   mPmText = "PM"
|   mPmTextView = androidx.appcompat.widget.AppCompatTextView@322364080 (0x1336e2b0)
|   mSecondPickerDescription = "Seconds circular slider"
|   mSecondSpaceView = androidx.appcompat.widget.AppCompatTextView@322364920 (0x1336e5f8)
|   mSecondView = com.wdullaer.materialdatetimepicker.AccessibleTextView@322365760 (0x1336e940)
|   mSelectHours = "Select hours"
|   mSelectMinutes = "Select minutes"
|   mSelectSeconds = "Select seconds"
|   mSelectedColor = -1
|   mThemeDark = false
|   mThemeDarkChanged = false
|   mTimePicker = com.wdullaer.materialdatetimepicker.time.RadialPickerLayout@322366584 (0x1336ec78)
|   mTitle = ""
|   mTypedTimes = java.util.ArrayList@322367336 (0x1336ef68)
|   mUnselectedColor = 1996488703
|   mVersion = com.wdullaer.materialdatetimepicker.time.TimePickerDialog$Version@321117312 (0x1323dc80)
|   mVibrate = true
|   mBackStackId = -1
|   mCancelable = true
|   mDialog = null
|   mDismissed = true
|   mShownByMe = false
|   mShowsDialog = true
|   mStyle = 0
|   mTheme = 0
|   mViewDestroyed = true
|   mAdded = false
|   mAnimationInfo = null
|   mArguments = null
|   mBackStackNesting = 0
|   mCalled = true
|   mChildFragmentManager = null
|   mContainer = null
|   mContainerId = 0
|   mDeferStart = false
|   mDetached = false
|   mFragmentId = 0
|   mFragmentManager = null
|   mFromLayout = false
|   mHasMenu = false
|   mHidden = false
|   mHiddenChanged = false
|   mHost = null
|   mInLayout = false
|   mInnerView = null
|   mIsCreated = false
|   mIsNewlyAdded = false
|   mLayoutInflater = null
|   mLifecycleRegistry = androidx.lifecycle.LifecycleRegistry@322367360 (0x1336ef80)
|   mMenuVisible = true
|   mParentFragment = null
|   mPerformedCreateView = false
|   mPostponedAlpha = 0.0
|   mRemoving = false
|   mRestored = false
|   mRetainInstance = false
|   mRetainInstanceChangedWhileDetached = false
|   mSavedFragmentState = null
|   mSavedStateRegistry = androidx.savedstate.bundle.BundleSavedStateRegistry@322367392 (0x1336efa0)
|   mSavedUserVisibleHint = null
|   mSavedViewState = android.util.SparseArray@322367416 (0x1336efb8)
|   mState = 0
|   mTag = null
|   mTarget = null
|   mTargetRequestCode = 0
|   mTargetWho = null
|   mUserVisibleHint = true
|   mView = null
|   mViewLifecycleOwner = null
|   mViewLifecycleOwnerLiveData = androidx.lifecycle.MutableLiveData@322367440 (0x1336efd0)
|   mWho = "abc9d6b6-647c-4fe0-b148-528a6c5369b2"
|   shadow$_klass_ = com.wdullaer.materialdatetimepicker.time.TimePickerDialog
|   shadow$_monitor_ = -1928200850
* Excluded Refs:
| Field: android.os.Message.obj
| Field: android.os.Message.next
| Field: android.os.Message.target
| Field: android.view.Choreographer$FrameDisplayEventReceiver.mMessageQueue (always)
| Field: android.view.ViewGroup$ViewLocationHolder.mRoot
| Thread:FinalizerWatchdogDaemon (always)
| Thread:main (always)
| Thread:LeakCanary-Heap-Dump (always)
| Class:java.lang.ref.WeakReference (always)
| Class:java.lang.ref.SoftReference (always)
| Class:java.lang.ref.PhantomReference (always)
| Class:java.lang.ref.Finalizer (always)
| Class:java.lang.ref.FinalizerReference (always)
```

**Comment by wdullaer** [comment]
I'm not sure I can fix this one in the library.
TimePickerDialog is the "root" class. It's the object you create in your app. I'm guessing you're holding on to a reference to the Dialog yourself.

Based on the log, a message you're dispatching from a worker thread has a reference to the TimePickerDialog.

If you look at the trace in the log, it doesn't really go "into" this library like the log in #586 does.

**Comment by archie94** [comment]
Yeah. Unable to find similar leaks in the sample app too. Closing this for now!

When can we expect to have the next release with 0a8fe28 ?

Thanks for all your support! :)

**Comment by wdullaer** [comment]
I've published 4.2.0 which includes that memory leak fix.
